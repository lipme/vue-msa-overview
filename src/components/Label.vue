<template>
  <g
    :class="classAttribute"
    :transform="transformAttribute"
  >
    <rect
      v-if="isSurrounded"
      :fill="background"
      :fill-opacity="rectFillOpacity"
      :x="rectX"
      :y="rectY"
      :height="fontSize+rectMargin"
      :stroke-width="rectBorderWidth"
      :stroke="borderColor"
      :width="rectWidth"
    >
     <animateTransform attributeName="transform"
             type="scale"
             from="0 0"
             to="1 1"
             begin="0s"
             dur="1s"
        />
    </rect>
    <text
      ref="text"
      :dx="optionsLabel.x"
      :dy="optionsLabel.y"
      :text-anchor="optionsLabel['text-anchor']"
      :font-size="optionsLabel['font-size']"
      :transform="optionsLabel.transform"
      :fill="labelColor"
    >
      {{this.label}}
      <animateTransform attributeName="transform"
             type="scale"
             from="0 0"
             to="1 1"
             begin="0s"
             dur="1s"
        />
    </text>
  </g>
</template>

<script>
export default {
  name: 'Label',
  props: {
    size: {
      type: Number,
      default: 6
    },
    x: {
      type: Number,
      default: 0
    },
    y: {
      type: Number,
      default: 0
    },
    label: {
      type: String,
      default: null
    },
    circular: {
      type: Boolean,
      default: false
    },
    id: {
      type: String,
      default: null
    },
    selected: {
      type: Boolean,
      default: false
    },
    color: {
      type: String,
      default: 'black'
    },
    background: {
      type: String,
      default: ''
    },
    borderWidth: {
      type: Number,
      default: 0
    },
    borderColor: {
      type: String,
      default: 'black'
    }
  },
  data () {
    return {
      rectWidth: 0
    }
  },
  mounted () {
    this.updateRect()
  },
  updated () {
    this.updateRect()
  },

  computed: {
    fontSize () {
      return this.size * 1.5
    },
    textClass () {
      if (this.circular && this.x > 180) {
        return 'reverse'
      }
      return 'normal'
    },
    classAttribute () {
      return 'nodeLabel ' + this.textClass + (this.selected ? ' selected' : '')
    },
    transformAttribute () {
      if (this.circular === false) {
        return 'translate(' + this.y + ' ' + this.x + ')'
      } else {
        return 'rotate(' + (this.x - 90) + ')translate(' + this.y + ')'
      }
    },
    optionsLabel () {
      // const transform = ''
      let anchor = 'start'
      let x = 1.5 * this.size

      if (this.circular && this.x > 180) {
        anchor = 'end'
        x = -1.5 * this.size
      }

      return {
        x: x,
        y: 0 + this.size / 2,
        'text-anchor': anchor,
        'font-size': this.fontSize + 'px'
      }
    },
    labelColor () {
      return this.selected ? 'red' : this.color
    },
    isSurrounded () {
      return this.background !== '' || this.rectBorderWidth > 0
    },
    rectMargin () {
      return this.fontSize / 5
    },
    rectBorderWidth () {
      return this.fontSize / 10 * this.borderWidth
    },
    rectY () {
      return this.optionsLabel.y - this.fontSize + this.rectMargin / 2
    },
    rectX () {
      // const margin = (this.circular && this.x > 180) ? this.optionsLabel.x - this.rectMargin * 3 : this.rectMargin * 2
      // return this.optionsLabel.x - margin
      return (this.circular && this.x > 180) ? this.rectMargin * 3 : this.optionsLabel.x - this.rectMargin * 2
    },
    rectFillOpacity () {
      return this.background ? '100%' : '0%'
    }

  },
  methods: {
    updateRect () {
      this.rectWidth = this.$refs.text.getComputedTextLength() + this.size * 1.2
    }
  }
}
</script>

<style scoped>
.selected text {
  font-weight: bold;
  transform: scale(1.5);
}

.selected rect {
  transform: scale(1.5);
}

.nodeLabel:hover text {
  transition: all 0.5s;
  transform: scale(1.5);
}

.nodeLabel:hover rect {
  transition: all 0.5s;
  transform: scale(1.5);
}

.reverse:hover text {
  transition: all 0.5s;
  transform: scale(1.5) rotate(180deg);
}

.reverse text {
  transform: rotate(180deg);
}

text {
  font-family: "Helvetica Neue, Helvetica, sans-serif";
}
</style>
