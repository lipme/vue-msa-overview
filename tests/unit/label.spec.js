import { shallowMount } from '@vue/test-utils'
import Label from '@/components/Label.vue'

describe('Label component', () => {
  it('default label renders correctly', () => {
    const comp = shallowMount(Label, {
      propsData: {
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('size set correctly', () => {
    const comp = shallowMount(Label, {
      propsData: {
        size: 1
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('x set correctly', () => {
    const comp = shallowMount(Label, {
      propsData: {
        x: 1
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('y set correctly', () => {
    const comp = shallowMount(Label, {
      propsData: {
        y: 1
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('label set correctly', () => {
    const comp = shallowMount(Label, {
      propsData: {
        label: 'label'
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('circular mode set correctly', () => {
    const comp = shallowMount(Label, {
      propsData: {
        circular: true
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('color set correctly', () => {
    const comp = shallowMount(Label, {
      propsData: {
        color: 'red'
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('background not set when length label is null', () => {
    const comp = shallowMount(Label, {
      propsData: {
        background: 'red'
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('background set correctly', () => {
    const comp = shallowMount(Label, {
      propsData: {
        background: 'red',
        label: 'label'
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('borderWidth has no effect when no label', () => {
    const comp = shallowMount(Label, {
      propsData: {
        borderWidth: 10
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('borderWidth set correctly', () => {
    const comp = shallowMount(Label, {
      propsData: {
        borderWidth: 10,
        label: 'label'
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('borderColor has no effect when no label', () => {
    const comp = shallowMount(Label, {
      propsData: {
        borderColor: 'red'
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('borderColor set correctly', () => {
    const comp = shallowMount(Label, {
      propsData: {
        borderColor: 'red',
        borderWidth: 10,
        label: 'label'
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('font-weight set correctly', () => {
    const comp = shallowMount(Label, {
      propsData: {
        fontWeight: 'bold'
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
})
