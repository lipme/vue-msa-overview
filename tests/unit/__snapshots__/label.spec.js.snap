// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label component background not set when length label is null 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 0)"
>
  <!---->
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component background set correctly 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 0)"
>
  <rect
    fill="red"
    fill-opacity="100%"
    height="10.8"
    stroke="black"
    stroke-width="0"
    width="29.7"
    x="5.4"
    y="-5.1"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </rect>
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    label
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component borderColor has no effect when no label 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 0)"
>
  <!---->
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component borderColor set correctly 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 0)"
>
  <rect
    fill=""
    fill-opacity="0%"
    height="10.8"
    stroke="red"
    stroke-width="9"
    width="38.7"
    x="5.4"
    y="-5.1"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </rect>
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    label
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component borderWidth has no effect when no label 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 0)"
>
  <!---->
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component borderWidth set correctly 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 0)"
>
  <rect
    fill=""
    fill-opacity="0%"
    height="10.8"
    stroke="black"
    stroke-width="9"
    width="38.7"
    x="5.4"
    y="-5.1"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </rect>
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    label
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component circular mode set correctly 1`] = `
<g
  class="nodeLabel normal"
  transform="rotate(-90)translate(0)"
>
  <!---->
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component color set correctly 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 0)"
>
  <!---->
   
  <text
    dx="9"
    dy="3"
    fill="red"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component default label renders correctly 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 0)"
>
  <!---->
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component font-weight set correctly 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 0)"
>
  <!---->
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="bold"
    text-anchor="start"
  >
    
    
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component label set correctly 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 0)"
>
  <!---->
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    label
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component size set correctly 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 0)"
>
  <!---->
   
  <text
    dx="1.5"
    dy="0.5"
    fill="black"
    font-size="1.5px"
    font-weight="normal"
    text-anchor="start"
  >
    
    
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component x set correctly 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(0 1)"
>
  <!---->
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;

exports[`Label component y set correctly 1`] = `
<g
  class="nodeLabel normal"
  transform="translate(1 0)"
>
  <!---->
   
  <text
    dx="9"
    dy="3"
    fill="black"
    font-size="9px"
    font-weight="normal"
    text-anchor="start"
  >
    
    
    
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </text>
</g>
`;
